components/calendar/src/astronomy.rs

```diff
@@ -0,0 +1,572 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+/// This file contains important structs and functions relating to location,
+/// time, and astronomy; these are intended for calender calculations and based off
+/// _Calendrical Calculations_ by Reingold & Dershowitz.
+use crate::iso::Iso;
+use crate::rata_die::RataDie;
+use crate::types::Moment;
+use crate::Date;
+
+pub struct Location {
+    latitude: f32,  // latitude from -90 to 90
+    longitude: f32, // longitude from -180 to 180
+    elevation: f32, // elevation in meters
+}
+
+// The mean synodic month in days of 86400 atomic seconds
+// (86400 = 24 * 60 * 60)
+const MEAN_SYNODIC_MONTH: f64 = 29.530588853;
+
+// The Moment of noon on January 1, 2000
+const J2000: Moment = Moment::new(730120.5);
+
+impl Location {
+    /// Create a location; latitude is from -90 to 90, and longitude is from -180 to 180
+    pub fn new(latitude_input: f32, longitude_input: f32, elevation_input: f32) -> Location {
+        let result = Location {
+            latitude: latitude_input,
+            longitude: longitude_input,
+            elevation: elevation_input,
+        };
+        result.check();
+        result
+    }
+
+    pub fn check(&self) {
+        if self.latitude < -90.0 || self.latitude > 90.0 {
+            debug_assert!(false, "Locations must have latitudes from -90.0 to 90.0.");
+        }
+        if self.longitude < -180.00 || self.longitude > 180.0 {
+            debug_assert!(false, "Locations must have longitudes from -180 to 180.");
+        }
+    }
+
+    /// Get the direction from loc1 to loc2 in degrees east of due north
+    pub fn direction(loc1: Location, loc2: Location) -> f32 {
+        let y = (loc2.longitude - loc1.longitude).to_radians().sin();
+        let x = (loc1.latitude.to_radians().cos() * loc2.latitude.to_radians().tan())
+            - (loc1.latitude.to_radians().sin()
+                * (loc1.longitude - loc2.longitude).to_radians().cos());
+        if (x == y && y == 0.0) || loc2.latitude == 90.0 {
```

Comment:
Note: it is error-prone to compare floating point numbers with exact equality; `0.0` _might_ be okay, and it _might_ be okay to compare against `loc2.latitude` since that is an exact value from a field, but you should in general avoid `==` comparisons with floating point values that are the result of any calculation.